src/shared/utilities/tsUtils.ts

```diff
@@ -31,3 +31,24 @@ type Expand<T> = T extends infer O ? { [K in keyof O]+?: O[K] } : never
  *
  */
 export type ExpandWithObject<T> = Expand<T> extends Record<string, unknown> ? Expand<T> : never
+
+/**
+ * Given two types, this yields a type that includes only the fields where both types were the exact same
+ * This is _almost_ equivalent to (T1 | T2) & T2 & T2, except that this type is distributive
+ */
+export type SharedTypes<T1, T2> = {
+    [P in keyof T1 & keyof T2]: (T1[P] | T2[P]) & T1[P] & T2[P]
+}
+
+/* All of the string keys of the shared type */
+export type SharedProp<T1, T2> = string & keyof SharedTypes<T1, T2>
+
+/* Any key that can be accumulated (i.e. an array) */
+export type AccumulatableKeys<T> = NonNullable<
+    {
+        [P in keyof T]: NonNullable<T[P]> extends any[] ? P : never
+    }[keyof T]
+>
+
+/** Similar to the nullish coalescing operator, but for types that can never occur */
+export type NeverCoalesce<T, U> = [T] extends [never] ? U : T
```

Comment:
could name it `Coalesce` ,

- avoids ambiguity: "Never Coalesce" sounds like it never coalesces :)
- has precedent in [sql](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/coalesce-transact-sql?view=sql-server-ver15)